README.md
# All Vibrate Package

The `all_vibrate` class provides methods for invoking a method channel to control phone vibrations in android.

## Getting Started

1. Add `all_vibrate` to the dependencies section of pubspec.yaml.

2.Import package

```

import 'package:all_vibrate/all_vibrate.dart';

```

## Methods

#### simpleVibrate

This method communicates with the native platform to trigger a vibration.

- Parameters:
  - period: The duration of the vibration in milliseconds.
  - amplitude: The amplitude of the vibration. Should be an integer value.

```

    await allVibrate.simpleVibrate(period: 500, amplitude: 50);

```

This will trigger a vibration for 500 milliseconds with an amplitude of 50
...

#### waveform

This method is used to trigger a custom vibration.

- Parameters:
  - timimgs: The duration of the vibration in milliseconds.
  - amplitude: The amplitude of the vibration. Should be an integer value.

```

    allVibrate.waveForm(timings: [200, 300, 400], amplitudes: [50, 75, 100]);

```

Vibrate 200ms with amplitude 50, next 300ms with amplitude 75 and so on

### Android

No configuration needed

### IOS

Coming soon


CHANGELOG.md
## 0.0.2

- minor changes
- code reformatted

## 0.0.1

* Initial release.
  -Implement simple vibration
  -Vibration with waveform


test/all_vibrate_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:all_vibrate/all_vibrate_platform_interface.dart';
import 'package:all_vibrate/all_vibrate_method_channel.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

class MockAllVibratePlatform
    with MockPlatformInterfaceMixin
    implements AllVibratePlatform {
  @override
  Future<void> simpleVibrate({required int period, required int amplitude}) {
    throw UnimplementedError();
  }

  @override
  Future<void> waveForm({
    required List<int> timings,
    required List<int> amplitudes,
  }) {
    throw UnimplementedError();
  }
}

void main() {
  final AllVibratePlatform initialPlatform = AllVibratePlatform.instance;

  test('$MethodChannelAllVibrate is the default instance', () {
    expect(initialPlatform, isInstanceOf<MethodChannelAllVibrate>());
  });

  test('getPlatformVersion', () async {
    MockAllVibratePlatform fakePlatform = MockAllVibratePlatform();
    AllVibratePlatform.instance = fakePlatform;
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:all_vibrate_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/all_vibrate_method_channel_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  const MethodChannel channel = MethodChannel('all_vibrate');

  setUp(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(
      channel,
      (MethodCall methodCall) async {
        return '42';
      },
    );
  });

  tearDown(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(channel, null);
  });
}


example/lib/main.dart
import 'package:all_vibrate/all_vibrate.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    final vibrate = AllVibrate();
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              ElevatedButton(
                child: const Text('Simple vibrate'),
                onPressed: () {
                  vibrate.simpleVibrate(period: 1500, amplitude: 200);
                },
              ),
              ElevatedButton(
                onPressed: () {
                  vibrate.waveForm(
                    timings: [102, 303, 983, 332],
                    amplitudes: [100, 150, 49, 100],
                  );
                },
                child: const Text('Waveform'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/all_vibrate_method_channel.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

import 'all_vibrate_platform_interface.dart';

/// An implementation of [AllVibratePlatform] that uses method channels.
class MethodChannelAllVibrate extends AllVibratePlatform {
  /// The method channel used to interact with the native platform.
  @visibleForTesting
  final methodChannel = const MethodChannel('all_vibrate');

  @override
  Future<void> simpleVibrate({
    required int period,
    required int amplitude,
  }) async {
    return await methodChannel.invokeMethod(
      "simple_vibrate",
      {
        "milliseconds": period,
        "amplitude": amplitude,
      },
    );
  }

  @override
  Future<void> waveForm({
    required List<int> timings,
    required List<int> amplitudes,
  }) async {
    if (timings.any((item) => item < 0) || amplitudes.any((item) => item < 0)) {
      throw Exception("Negative number not accepted");
    }
    return await methodChannel.invokeMethod(
      "wave_form",
      {
        "timings": timings,
        "amplitudes": amplitudes,
      },
    );
  }
}


lib/all_vibrate_platform_interface.dart
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

import 'all_vibrate_method_channel.dart';

abstract class AllVibratePlatform extends PlatformInterface {
  /// Constructs a AllVibratePlatform.
  AllVibratePlatform() : super(token: _token);

  static final Object _token = Object();

  static AllVibratePlatform _instance = MethodChannelAllVibrate();

  /// The default instance of [AllVibratePlatform] to use.
  ///
  /// Defaults to [MethodChannelAllVibrate].
  static AllVibratePlatform get instance => _instance;

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [AllVibratePlatform] when
  /// they register themselves.
  static set instance(AllVibratePlatform instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  Future<void> simpleVibrate({
    required int period,
    required int amplitude,
  });

  Future<void> waveForm({
    required List<int> timings,
    required List<int> amplitudes,
  });
}


android/src/main/kotlin/com/rans_innovations/all_vibrate/AllVibratePlugin.kt
@file:Suppress("DEPRECATION")

package com.rans_innovations.all_vibrate

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** AllVibratePlugin */
class AllVibratePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private var manager: VibratorManager? = null
    private lateinit var vibrator: Vibrator
    private lateinit var ctx: Context


    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "all_vibrate")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {


        manager = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {

            ctx.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
        } else {
            null
        }
        vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            manager!!.defaultVibrator
        } else {
            ctx.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }

        val args = call.arguments as Map<*, *>

        when (call.method) {
            "simple_vibrate" -> {
                val milliseconds = args["milliseconds"] as Int
                val amplitude = args["amplitude"] as Int
                vibrator.cancel()
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    vibrator.vibrate(
                        VibrationEffect.createOneShot(
                            milliseconds.toLong(),
                            amplitude
                        )
                    )
                } else {
                    vibrator.vibrate(milliseconds.toLong())
                }
            }

            "wave_form" -> {
                vibrator.cancel()
                val ticks = args["timings"] as List<*>
                val amps = args["amplitudes"] as List<*>


                val timings =
                    convertToLongArray(ticks)
                val amplitude = convertToIntArray(amps)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    vibrator.vibrate(VibrationEffect.createWaveform(timings, amplitude, -1))
                } else {
                    vibrator.vibrate(timings, -1)
                }
            }
        }
    }

    private fun convertToLongArray(item: List<*>): LongArray {
        val longArray = LongArray(item.size)
        item.forEachIndexed { index, it ->
            longArray[index] =
                it.toString().toLong()
        }
        return longArray
    }

    private fun convertToIntArray(item: List<*>): IntArray {
        val intArray = IntArray(item.size)
        item.forEachIndexed { index, it ->
            intArray[index] = it.toString().toInt()
        }
        return intArray
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        ctx = binding.activity.applicationContext
    }

    override fun onDetachedFromActivityForConfigChanges() {
        TODO("Not yet implemented")
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        TODO("Not yet implemented")
    }

    override fun onDetachedFromActivity() {
        TODO("Not yet implemented")
    }
}


lib/all_vibrate.dart
import 'all_vibrate_platform_interface.dart';

/// The `AllVibrate` class provides methods to control phone vibrations.
///
/// Example:
/// ```dart
/// // Create an instance of AllVibrate.
/// var allVibrate = AllVibrate();
///
/// // Vibrate the phone for 500 milliseconds with an amplitude of 50.
/// await allVibrate.simpleVibrate(period: 500, amplitude: 50);
///
/// // Vibrate the phone with a specified waveform pattern.
/// await allVibrate.waveForm(timings: [200, 300, 400], amplitudes: [50, 75, 100]);
/// ```
class AllVibrate {
  /// Invokes a method channel to vibrate the phone with a specified duration and amplitude.
  ///
  /// This method communicates with the native platform to trigger a vibration.
  ///
  /// - Parameters:
  ///   - period: The duration of the vibration in milliseconds.
  ///   - amplitude: The amplitude of the vibration. Should be an integer value.
  ///
  /// - Returns: A [Future] that completes when the vibration is initiated.
  ///
  /// Example:
  /// ```dart
  /// // Vibrate the phone for 500 milliseconds with an amplitude of 50.
  /// await allVibrate.simpleVibrate(period: 500, amplitude: 50);
  /// ```
  Future<void> simpleVibrate({
    required int period,
    required int amplitude,
  }) {
    return AllVibratePlatform.instance.simpleVibrate(
      amplitude: amplitude,
      period: period,
    );
  }

  /// Invokes a method channel to vibrate the phone with a specified waveform pattern.
  ///
  /// This method communicates with the native platform to trigger a vibration pattern
  /// based on the provided timings and amplitudes. The lengths of the timings and
  /// amplitudes lists must be equal, or an exception will be thrown.
  ///
  /// - Parameters:
  ///   - timings: A list of integers representing the duration of each vibration segment
  ///     in milliseconds.
  ///   - amplitudes: A list of integers representing the amplitude of each respective
  ///     vibration segment. The length of this list must match the length of the timings list.
  ///
  /// - Returns: A [Future] that completes when the waveform vibration is initiated.
  ///
  /// Example:
  /// ```dart
  /// await allVibrate.waveForm(timings: [200, 300, 400], amplitudes: [50, 75, 100]);
  /// ```
  Future<void> waveForm({
    required List<int> timings,
    required List<int> amplitudes,
  }) {
    if (timings.length != amplitudes.length) {
      throw ArgumentError(
          '''Lengths of timings and amplitudes lists must be equal.
      Length of timings is ${timings.length} and length of amplitudes is ${amplitudes.length}
      ''');
    }

    return AllVibratePlatform.instance.waveForm(
      amplitudes: amplitudes,
      timings: timings,
    );
  }
}
